xds: deflake Test/ServerSideXDS_ServingModeChanges

easwars opened the pull request:
The test creates two listeners, and both are in `serving` mode initially. We also have two clientConns to the two servers (backed by those two listeners), and these are initially in `READY` state and are able to make a successful RPC. 

Then we remove one of the listener resources, which gets one of the servers into `not-serving` mode. At this point, we check if one of the clientConns has moved out of `READY` (because the server is not serving anymore) and the other one remains in `READY` (since this server is not affected and is still serving).

Currently, grpc-go does not suppress xds resource updates which change only in the *version* number. Therefore, the test receives a `Listener` update even for the resource which has not been updated, and since we updated the A36 spec recently to say the following:
```
Updates to a Listener cause all older connections on that Listener to be
gracefully shut down with a grace period of 10 minutes for long-lived
RPC's, such that clients will reconnect and have the updated
configuration apply.
```
we drain existing transports, which moves the `ClientConn` to `IDLE`. 

This PR fixes the test to rely solely on the outcome of RPCs and not on the channel state.

RELEASE NOTES: N/A

dfawley commented:
This doesn't seem right.  We shouldn't expect the client to go into IDLE so quickly.  I'll take a look at this since you are out today & I made the IDLE change.

dfawley commented:
Okay, I think I see what's happening.  When we send this update here:

https://github.com/grpc/grpc-go/blob/8ab16ef276a33df4cdb106446eeff40ff56a6928/xds/internal/test/xds_server_serving_mode_test.go#L187-L190

Even though the listener resource is not modified, an update comes to the xds client and to the first `xds.GRPCServer`.  This causes the server to kill all its incoming connections.  This seems like a bug to me - if the listener in question doesn't change, we don't want to make all connections cycle.

You can make this failure happen reliably by adding a 2 second sleep after sending the update that removes the second listener.  I also locally deleted the second listener entirely from this test case, which may help with debugging (and unless there is a strong reason to have both intermingled in the same test, we may want to separate these test cases in the future).

easwars commented:
> Okay, I think I see what's happening. When we send this update here:
> 
> https://github.com/grpc/grpc-go/blob/8ab16ef276a33df4cdb106446eeff40ff56a6928/xds/internal/test/xds_server_serving_mode_test.go#L187-L190
> 
> Even though the listener resource is not modified, an update comes to the xds client and to the first `xds.GRPCServer`. This causes the server to kill all its incoming connections. This seems like a bug to me - if the listener in question doesn't change, we don't want to make all connections cycle.
> 
> You can make this failure happen reliably by adding a 2 second sleep after sending the update that removes the second listener. I also locally deleted the second listener entirely from this test case, which may help with debugging (and unless there is a strong reason to have both intermingled in the same test, we may want to separate these test cases in the future).

As discussed offline, we have changed the xds-enabled gRPC server implementation to drain all transports for every `Listener` resource update. Currently, the `xdsClient` does not filter out updates which change only in the version number. This is causing the server in the test to drain existing transports even when nothing has changed in the listener. I've changed this PR to only fix this test, and make it not flaky.

Filed https://github.com/grpc/grpc-go/issues/4695 to track the redundant resource issue.

dfawley commented on xds/internal/test/xds_server_serving_mode_test.go:
```diff
@@ -261,11 +243,11 @@ func (s) TestServerSideXDS_ServingModeChanges(t *testing.T) {
 	}
-	if _, err := client2.EmptyCall(ctx, &testpb.Empty{}, grpc.WaitForReady(true)); err != nil {
-		t.Fatalf("rpc EmptyCall() failed: %v", err)
+	if err := waitForSuccessfulRPC(ctx, cc2); err != nil {
+		t.Fatal(err)
 	}
```
If this is literally always a `t.Fatal`, maybe pass `t` there, `t.Helper` it, and save 2 LoC every time it's called?  (Same with `waitForFailedRPC`.)

easwars replied:
Done.
